### PR TITLE
[DOCS] Improve Type System docs

### PR DESCRIPTION
**Pull request by DSouzaM**
I made some improvements to the Type System docs while reading through them. Things improved/changed:

- general readability
- description of structural typing (I [don't think it's used in Haskell](https://www.google.ca/search?q=haskell+structural+typing), and in OCaml/Elm it seems to be [just used for record types](https://en.wikipedia.org/wiki/Structural_type_system#Description). Correct me if I've made a mistake)
- ~~description of soundness vs completeness (I think their meanings were reversed)~~

### CODE REVIEW

File: website/en/docs/lang/types-and-expressions.md [hunk 1]
```diff
@@ -69,30 +69,30 @@ of values. But still Flow checks _every_ possible value. In this way Flow is
...
-actually occur when the code is run. Flow does this in order to be _"sound"_.
+actually occur when the code is run. Flow does this in order to be _"complete"_.
 
-In type systems, ***soundness*** is the ability for a type checker to catch
+In type systems, ***completeness*** is the ability for a type checker to catch
 every single error that _might_ happen at runtime. This comes at the cost of
 sometimes catching errors that will not actually happen at runtime.
 
-On the flip-side, ***completeness*** is the ability for a type checker to only
+On the flip-side, ***soundness*** is the ability for a type checker to only
```
Comment (rattrayalex) [inline website/en/docs/lang/types-and-expressions.md]: Hmm, really? This is a pretty big change to my understanding.

Comment (DSouzaM) [reply website/en/docs/lang/types-and-expressions.md]: To my knowledge, a sound system will only ever make correct decisions (an unsound system might make incorrect decisions, e.g. letting program X pass type-checking despite having a type error). A complete system makes it possible, for any correct program, to show that it is correct (an incomplete system might not be sufficiently powerful to prove that program X is valid even if it is).

I'm drawing my definitions from [first order logic](https://math.stackexchange.com/a/105576), but maybe I'm misunderstanding it in this context.

Comment (asolove) [reply website/en/docs/lang/types-and-expressions.md]: I too think this was correct in the original. Soundness means it does not allow any incorrect program, but may also not allow programs that would be fine at run time. Completeness means that any correct program is allowed, although possibly some incorrect ones are as well.

Comment (DSouzaM) [reply website/en/docs/lang/types-and-expressions.md]: Okay, I follow now. Fixed.

Comment (clemmy) [reply website/en/docs/lang/types-and-expressions.md]: @asolove Your explanation seems correct to me. The part that makes the documentation confusing IMO is

> On the flip-side, completeness is the ability for a type checker to only ever catch errors that __would happen__ at runtime. This comes at the cost of sometimes missing errors that __will happen__ at runtime.

If there is a way to phrase this better, it'll be a lot more clear.

**Comment by SamChou19815** [comment]
I will close this since it has not been updated for years, and we just pushed out a new website so the edited files are no longer there.

**Pull request closed without merging**